### PR TITLE
Fix out-of-bounds accesses and modernize some line-number do loops

### LINKED ISSUE #558: Code reads out-of-bounds at LM+1in three places
While testing another change, I found these out-of-bounds accesses:

- Twice in MDL2SIGMA.f, the code can access element LM+1 if NL1XF(I,J)=LP1
- UPP_PHYSICS.f: model can access out of array bounds when LCL(I,J) is at the model top

These happened in an RRFS case with extreme atmospheric conditions.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
This fixes #558 and replaces three "do NUMBER ... NUMBER continue" loops with modern "do..enddo" loops.

### CODE REVIEW

File: sorc/ncep_post.fd/MDL2SIGMA.f [hunk 1]
```diff
@@ -277,8 +277,9 @@ SUBROUTINE MDL2SIGMA
            AHF  =D00
            FAC  =D00
 	   DONEFSL1=.TRUE.
-         ELSEIF(T(I,J,NL1XF(I,J))<SPVAL                 &
-     &	 .AND. Q(I,J,NL1XF(I,J))<SPVAL)THEN
+         ELSEIF(NL1XF(I,J)<LP1) THEN
+          IF(T(I,J,NL1XF(I,J))<SPVAL                 &
+     &	   .AND. Q(I,J,NL1XF(I,J))<SPVAL)THEN
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/MDL2SIGMA.f]: @SamuelTrahanNOAA Do you think two if statements can be combined?

File: sorc/ncep_post.fd/MDL2SIGMA.f [hunk 2]
```diff
@@ -656,8 +658,9 @@ SUBROUTINE MDL2SIGMA
 	      TSLDONE=.TRUE.
 !
 !
-            ELSEIF(T(I,J,NL1XF(I,J))<SPVAL              &
-     &	 .AND. Q(I,J,NL1XF(I,J))<SPVAL)THEN
+            ELSEIF(NL1XF(I,J)<LP1) THEN
+             IF(T(I,J,NL1XF(I,J))<SPVAL              &
+     &	  .AND. Q(I,J,NL1XF(I,J))<SPVAL)THEN
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/MDL2SIGMA.f]: @SamuelTrahanNOAA Do you think two if statements can be combined?

File: sorc/ncep_post.fd/MDL2SIGMA.f (hunk 1 above)
Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/MDL2SIGMA.f]: @SamuelTrahanNOAA Please add a change log with your name at the beginning session of this subroutine.

File: sorc/ncep_post.fd/UPP_PHYSICS.f [hunk 3]
```diff
@@ -1320,6 +1320,9 @@ SUBROUTINE CALCAPE2(ITYPE,DPBND,P1D,T1D,Q1D,L1D,    &
               THUNDER(I,J) = .FALSE.
             ENDIF
           ENDIF
+
+          ! Limit LCL to prevent out-of-bounds accesses later
+          LCL(I,J) = max(min(LCL(I,J),LM-1),1)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/UPP_PHYSICS.f:1326]: @SamuelTrahanNOAA Please add a change log with your name at the beginning session of this subroutine.

**Comment by WenMeng-NOAA** [comment]
The UPP RT test was completed on Hera. No baseline changes are needed.

File: sorc/ncep_post.fd/MDL2SIGMA.f (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/MDL2SIGMA.f]: No, they cannot be combined. Fortran has no short-circuit "and" operators. The `NL1XF(I,J)<LP1` must be checked before that index of an array is accessed. The only way to do that in Fortran is a nested "if" statement.

File: sorc/ncep_post.fd/MDL2SIGMA.f (hunk 2 above)
Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/MDL2SIGMA.f]: No, they cannot be combined. Fortran has no short-circuit "and" operators. The `NL1XF(I,J)<LP1` must be checked before that index of an array is accessed. The only way to do that in Fortran is a nested "if" statement.

File: sorc/ncep_post.fd/UPP_PHYSICS.f (hunk 3 above)
Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/UPP_PHYSICS.f:1326]: Done. I added one to the other file too.

File: sorc/ncep_post.fd/MDL2SIGMA.f (hunk 2 above)
Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/MDL2SIGMA.f]: I've added a comment to this effect.

File: sorc/ncep_post.fd/MDL2SIGMA.f (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/MDL2SIGMA.f]: I've added a comment to this effect.

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/MDL2SIGMA.f]: @SamuelTrahanNOAA I see it. Thanks!

**Comment by WenMeng-NOAA** [comment]
The UPP RT test was completed on WCOSS2.

**Review by WenMeng-NOAA — APPROVED** [review]

**Pull request merged**